### ISSUE TITLE
Synchronize Batch Norm across Multi GPUs

### ISSUE THREAD

**Issue by ycszen** [pytorch/pytorch#2584]
I find in some tasks , for example, semantic segmentation, detection,  sync batch norm is crucial for performance.In these tasks, batch size per gpu is small so we need sync the bn's mean and variance   to enlarge the bn's "batch size". It need synchronize parameters and gradient across multi gpus in forward and backward. This will obviously improve the performance. I believe this will attract more and more users. So can you implement it in pytorch?

**Comment by ngimel** [comment]
In many applications, having per-GPU batchnorm is both acceptable and desirable, (see, e.g. https://arxiv.org/pdf/1706.02677.pdf), so multi-GPU batch norm should be optional. Parameters and gradients are already synchronized in DataParallelTable, mean and variance, however, are not. Torch batchnorm implementation would have to be rewritten to use stable single pass variance computation, but still, this will introduce a sync point across the GPUs, so it will have performance (as in, speed of execution, not model accuracy) impact.

**Comment by fmassa** [comment]
@ycszen What you could do is to fix the batch-norm statistics in this case, or even (and it is what most people do) replace entirely batch-norm with a fixed affine transformation.

**Comment by ycszen** [comment]
@ngimel Yes. It is optional. In recognition task, the batch size per gpu is large, so this is not necessary. However, in semantic segmentation or detection, the batch size per gpu is so small, even one image per gpu, so the multi-GPU batch norm is crucial. Even in some case, we cannot reproduce the performance in the paper without multi-GPU, for example PSPNet or Deeplab v3.
Personally, we can offer a parameter to control using multi-GPU or not.

**Comment by ycszen** [comment]
@fmassa My opinion is that in the forward we can gather the multi-GPU batchnorm's input, then synchronize them to compute the statistics while in the backward we synchronize the gradient. This can enlarge the batchnorm's batchsize, so it can compute more precise statistics. This is why it can improve performance in some tasks which have small batch size. How about your opinion?

**Comment by sunalbert** [comment]
@ycszen I can't agree with you any more. In practice, I find sync batch norm is very useful.

**Comment by ycszen** [comment]
@sunalbert Yes.In practice, this is very useful, so I want the authors can implement it for the open community.Or give me some guides, I can implement it by myself.

**Comment by Eniac-Xie** [comment]
@ycszen Agree with you

**Comment by amdcat** [comment]
Yes, it's a important problem when use small batch size. 
Is there a solution in pytorch?

**Comment by Chillee** [comment]
This implementation might be of interest to people: http://hangzh.com/PyTorch-Encoding/syncbn.html

**Comment by CanyonWind** [comment]
@Chillee  Nice work!

**Comment by zhanghang1989** [comment]
Thanks @CanyonWind . I am collecting the credits :)

**Comment by Chillee** [comment]
@zhanghang1989 @CanyonWind Yep, definitely not my work. All credit goes to @zhanghang1989 !

**Comment by mario98** [comment]
I really think if there was a synchronized batch norm as an official part of the framework maintained by the pytorch developers, this would be of great value and attract a lot of people to use pytorch. As already said, in tasks like semantic segmentation or object detection, model size is too big for a reasonable batch size per GPU, but to compete against state of the art who all use synced batch norm or some other tricks (like in-place batch norm, but that's another story) you have to do a lot of implementations yourself since they usually keep their implementation secret (like PSPNet) because they know how key this is to high accuracies.
While I appreciate all attempts made by others to provide sync batch norm as an unofficial extension, my experience is that none of these implementations are stable enough for people to use it and they often suffer from compatibility issues since pytorch is developed with the gloves off and thinks get broken from version to version (sometimes even undocumented). I requires a lot of tinkering to get those to run and even then unexpected behavior happens again and again.

On the other side, if pytorch developers fail to provide such a functionality, and another framework steps in and provides synced batch norm out-of-the-box, pytorch will be over the hill. The other framework will win. Legend says that approximately each year, a new disrupting framework emerges and the old ones are unable to compete any more. Remember torch, caffe, theano? (and maybe soon tensorflow? Although tensorflow recognized the sign of the times and implemented dynamic graphs, so chances are that tensorflow will survive, at least for now...)
Sync batch norm is for many developers the killer feature. And pytorch already had a great year in 2017... I'm just saying...

**Comment by kmaninis** [comment]
Hi, [this](http://hangzh.com/PyTorch-Encoding/nn.html#batchnorm2d) is the updated link from above. It works quite well! Needs installation of yet another external package, though. It would be great if incorporated in the pytorch repo.

**Comment by zhanghang1989** [comment]
FYI, I have implemented SyncBatchNorm using MXNet Gluon http://mxnet.incubator.apache.org/versions/master/api/python/gluon/contrib.html?highlight=syncbatchnorm#mxnet.gluon.contrib.nn.SyncBatchNorm

**Comment by soumith** [comment]
Added now in PyTorch master via https://github.com/pytorch/pytorch/pull/14267
For documentation, see: https://pytorch.org/docs/master/nn.html#torch.nn.SyncBatchNorm

**Comment by dougsouza** [comment]
Is there any chance that we could use this with DataParallel?

**Comment by soumith** [comment]
@dougsouza unfortunately no

**Comment by dougsouza** [comment]
For those who are struggling to get DistributedDataParallel working to use SyncBatchnorm, I wrote a small step-by-step [here](https://github.com/dougsouza/pytorch-sync-batchnorm-example). Hope it helps
